vpr/src/base/vpr_constraints.h

```diff
@@ -0,0 +1,73 @@
+#ifndef VPR_CONSTRAINTS_H
+#define VPR_CONSTRAINTS_H
+
+#include "vtr_vector.h"
+#include "vpr_utils.h"
+#include "partition.h"
+#include "partition_region.h"
+
+/**
+ * @file
+ * @brief This file defines the VprConstraints class used to store and read out data related to user-specified
+ * 		  block and region constraints for the packing and placement stages.
+ *
+ * Overview
+ * ========
+ * This class contains functions that read in and store information from a constraints XML file.
+ * The XML file provides a partition list, with the names of the partitions, and each atom in the partition.
+ * It also specifies which regions the partitions should be placed in. Atoms cannot be placed in more than one partition.
+ * If an atom is assigned to more than one partition, the last partition is was assigned to will be the partition it is placed in.
+ *
+ */
+
+class VprConstraints {
+  public:
+    /**
+     * @brief Store the id of a constrained atom with the id of the partition it belongs to
+     *
+     *   @param blk_id      The atom being stored
+     *   @param part_id     The partition the atom is being constrained to
+     */
+    void add_constrained_atom(const AtomBlockId blk_id, const PartitionId part_id);
+
+    /**
+     * @brief Return id of the partition the atom belongs to
```

Comment:
// If an atom is not in a partition (unconstrained), PartitionId::Invalid() is returned.